gpii/node_modules/lifecycleManager/src/LifecycleManager.js

```diff
@@ -568,13 +587,79 @@ var gpii = fluid.registerNamespace("gpii");
+    /**
+     * Read corresponding setting values of the requested preferences. The read result is applied as
+     * a model change to userSession.model.currentPreferences.
+     * @param {Component} that - An instance of gpii.lifecycleManager.
+     * @param {Object} preference - The preference to read settings for. An example is:
+     * "http://registry\\.gpii\\.net/common/magnification": {
+     *     value: 3
+     * }
+     * @param {Event} preferenceReadSuccessEvent - The event fired when the read succeeds.
+     * @param {Event} preferenceReadFailEvent - The event fired when the read fails.
+     */
+    gpii.lifecycleManager.readPreference = function (that, preference, preferenceReadSuccessEvent, preferenceReadFailEvent) {
+        fluid.log("Lifecycle Manager: Received preference to read ", preference);
+        var session = that.getSession();
+
+        // Mock values for running preference thru matchMaker
+        var contextualPrefs = {};
+        var activeContextName = "gpii-default";
+        var prefSet = {};
+        fluid.each(preference, function (spec, pref) {
+            var prefPath = fluid.pathUtil.parseEL(pref);
+            fluid.set(contextualPrefs, prefPath, spec.value);
+        });
+        fluid.set(prefSet, ["contexts", activeContextName, "preferences"], contextualPrefs);
+
+        // Perform a local matchmaking to get matched solutions for the preference.
+        var matchMakingPromise = that.privateMatchMaker.doMatch("readSetting", prefSet);
+        matchMakingPromise.then(function (payload) {
+            fluid.set(payload, ["activeContextName"], activeContextName);
+            fluid.set(payload, ["preferences"], prefSet);
+            that.addLifecycleInstructionsToPayload(payload);
+
+            // compile a list of solution registry entries for all matched solutoins
+            var matchMakerSolutionRegistryEntries = fluid.get(payload, ["activeConfiguration", "lifecycleInstructions"]);
+            var settingsPromise = that.read(matchMakerSolutionRegistryEntries);
+            settingsPromise.then(function (allSettingValues) {
+                // As lifecycleManager.read() returns values for all settings that matched solutions support,
+                // only filter out setting values that are in the matchmaker result.
+                var matchMakerSettings = gpii.lifecycleManager.getMatchMakerSettings(allSettingValues, matchMakerSolutionRegistryEntries);
+
+                // Transform settings to preferences
+                var transformed = gpii.lifecycleManager.transformSettingsToPrefs(matchMakerSettings, matchMakerSolutionRegistryEntries);
+
+                // Filter matched preferences
+                var matchedPrefs = gpii.lifecycleManager.filterMatchedPrefs(transformed, contextualPrefs);
+
+                if (fluid.jQueryStandalone.isEmptyObject(matchedPrefs)) {
+                    preferenceReadFailEvent.fire();
```

Comment:
Is this really a failure? My impression is that the only real cause of a failure would be a promise rejection from the settings handler. It's unclear in what cases we would return an empty object here and what kind of failure this might be.

Reply:
As a general note I always seem to find any use of isEmptyObject is a possibly suspicious smell which is why it has been left hosted at the jQuery level : P

Reply:
Good catch on settings handler failures. I've improved the code to cover this failure.

> It's unclear in what cases we would return an empty object here and what kind of failure this might be.
Use cases discovered so far that will return an empty object here:
1. Settings handlers that have get() defined but not implemented - https://issues.gpii.net/browse/GPII-4217
2. "inverseCapabilitiesTransformations" blocks for one or more settings are not defined. In this case, although setting values are read successfully, when these values are transformed into preference values via gpii.lifecycleManager.transformSettingsToPrefs(), an empty object will return - https://issues.gpii.net/browse/GPII-4216

Overall, an empty object means a matched preference is not found and the read fails.